src/queue.rs

```diff
@@ -512,14 +512,63 @@ impl<M: GuestAddressSpace> Queue<M> {
         }
     }
 
+    /// Update avail_event on the used ring with the last index in the avail ring.
+    pub fn update_avail_event(&mut self) {
+        let mem = self.mem.memory();
+        let index_addr = match mem.checked_offset(self.avail_ring, 2) {
+            Some(ret) => ret,
+            None => {
+                // TODO log address
+                warn!("Invalid offset");
+                return;
+            }
+        };
+        // Note that last_index has no invalid values
+        let last_index: u16 = match mem.read_obj::<u16>(index_addr) {
+            Ok(ret) => ret,
+            Err(_) => return,
+        };
+
+        match mem.checked_offset(self.used_ring, (4 + self.actual_size() * 8) as usize) {
+            Some(a) => {
+                mem.write_obj(last_index, a).unwrap();
+            }
+            None => warn!("Can't update avail_event"),
+        }
+
+        // This fence ensures the guest sees the value we've just written.
+        fence(Ordering::Release);
```

Comment:
This fence is not needed. However, you need a SeqCst fence in the consumer, before it checks the `avail_idx`.

The way QEMU does it is to have a different API, `enable_notification(bool)`. The API does the same thing as your `update_avail_event` if `VIRTIO_RING_F_EVENT_IDX` is not available, but it could also takes care of setting or clearing `VRING_USED_F_NO_NOTIFY` (depending on the bool argument) if event indices are not in use.

With this API, the SeqCst fence is only needed if the bool argument is true.